### PR TITLE
TryExecuteQuery

### PR DESCRIPTION
**Pull request by bchong95**
# TryExecuteQuery

## Description

This PR adds support for a TryExecuteQuery method which will try to execute a query with a set of supported query features. If the query is supported than the method will return a QueryIterator. If the query is not supported than the method returns a QueryPlan instead. 

The purpose of this method is to support SQL Query Compute Gateway. We are going to initially only support queries that don't require any distribution work and over time allow more queries. If the query is not supported than we will send back a bad request with the query plan as usual and allow the SDK to do the cross partition work. The method will also always return a QueryPlan even if the query is executable, since we might want to cache the query plan in the future. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 1]
```diff
@@ -799,6 +800,13 @@ public abstract class Container
             string continuationToken = null,
             QueryRequestOptions requestOptions = null);
 
+        internal abstract Task<(PartitionedQueryExecutionInfo, (bool, QueryIterator))> TryExecuteQueryAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: Container only has the public contract. The internal methods are only in the ContainerCore.

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs [hunk 2]
```diff
@@ -247,6 +248,79 @@ internal partial class ContainerCore : Container
                 requestOptions: requestOptions);
         }
 
+        internal override async Task<(PartitionedQueryExecutionInfo, (bool, QueryIterator))> TryExecuteQueryAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Should this be in Friends?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs [hunk 3]
```diff
@@ -247,6 +248,79 @@ internal partial class ContainerCore : Container
...
+
+            if (requestOptions == null)
+            {
+                throw new ArgumentNullException(nameof(requestOptions));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            Documents.PartitionKeyDefinition partitionKeyDefinition;
+            if (requestOptions.Properties != null
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Is there any reason all this logic is not included in the Cosmos query factory class?

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: How will any user call the method without casting to ContainerCore then?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: I would like to but it takes two different changes, then how will I test it here?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs (hunk 3 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: You would have add a public method in CosmosQueryFactory for that to work. I also don't really get what this all is. The user should pass in a PartitionKey in the FeedOptions if they need one. Ideally it should just be in the query body and then we wouldn't need this code.

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: That is correct. That is the way it is currently model. We don't want to leak internal implementation into public contract.

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: Removed it from the public contract (even though it was already internal).

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: I am leaving as internal for now.

**Comment by sboshra** [comment]
>         public async Task<PartitionedQueryExecutionInfo> GetQueryPlanAsync(

please delete if not used anymore 

---
Refers to: Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs:30 in 602869d. [](commit_id = 602869d10d2c47e0d88007bf450d913f3d4de836, deletion_comment = False)

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs [hunk 4]
```diff
@@ -44,6 +46,55 @@ public QueryPlanHandler(CosmosQueryClient queryClient)
...
+
+            return partitionedQueryExecutionInfo;
+        }
+
+        public async Task<(PartitionedQueryExecutionInfo, bool)> GetQueryInfoAndIfSupportedAsync(
+            QueryFeatures supportedQueryFeatures,
+            SqlQuerySpec sqlQuerySpec,
+            PartitionKeyDefinition partitionKeyDefinition,
+            bool hasLogicalPartitionKey,
+            CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs]: cancellation

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs [hunk 5]
```diff
@@ -151,114 +148,173 @@ private static class QueryPlanExceptionFactory
...
+        {
+            private static readonly IReadOnlyList<QueryFeatures> QueryFeatureList = (QueryFeatures[])Enum.GetValues(typeof(QueryFeatures));
+            private static readonly ReadOnlyDictionary<QueryFeatures, ArgumentException> FeatureToUnsupportedException = new ReadOnlyDictionary<QueryFeatures, ArgumentException>(
+                QueryFeatureList
+                .ToDictionary(
+                    x => x,
+                    x => new ArgumentException(QueryPlanExceptionFactory.FormatExceptionMessage(x.ToString()))));
 
-            private static void AddExceptionsForTopQueries(
+            public static void ThrowIfNotSupported(
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs]: >ThrowIfNotSupported [](start = 31, length = 19)

check if we can delete this function

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs [hunk 6]
```diff
@@ -15,13 +15,27 @@ internal class QueryIterator : FeedIterator
         private readonly CosmosSerializationFormatOptions cosmosSerializationFormatOptions;
 
         internal QueryIterator(
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs]: >internal [](start = 8, length = 8)

private

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Is there anything blocking from writing unit tests or mocking out to do the tests? We only want SDK in the SDK. Everything else should be in friends.

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Lets please not pollute SDK with friends concerns.

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: The whole class is internal. I can remove the tests if that will get the PR unblocked.

**Comment by kirankumarkolli** [comment]
@bchong95 won;t it get solved by query wire protocol versioning and configuring SDK to use target version?

**Comment by bchong95** [comment]
There isn't any versioning or configuring here

**Comment by bchong95** [comment]
This is merely exposing a TryExecute variant of CreateDocumentQuery. They are calling into the same version of code.

**Comment by bchong95** [comment]
It also allows the caller to determine what queries they want to allow.

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs [hunk 7]
```diff
@@ -44,6 +46,57 @@ public QueryPlanHandler(CosmosQueryClient queryClient)
...
+                cancellationToken);
+
+            QueryPlanHandler.QueryPlanExceptionFactory.ThrowIfNotSupported(
+                partitionedQueryExecutionInfo.QueryInfo,
+                supportedQueryFeatures);
+
+            return partitionedQueryExecutionInfo;
+        }
+
+        public async Task<(PartitionedQueryExecutionInfo, bool)> GetQueryInfoAndIfSupportedAsync(
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs]: Is this the new query compute contract, If so isn't friends is the right place?

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 8]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.Azure.Cosmos
     using System.Linq;
     using System.Threading;
     using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Query;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: undo

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs (hunk 7 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/QueryPlanHandler.cs]: It requires private members so it's a no go.

**Comment by kirankumarkolli** [comment]
@bchong95 please refresh the description also with current state of PR.

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs (hunk 8 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs]: I ended up needing queryclient.

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs [hunk 9]
```diff
@@ -247,6 +248,82 @@ internal partial class ContainerCore : Container
...
+                && requestOptions.Properties.TryGetValue("x-ms-query-partitionkey-definition", out object partitionKeyDefinitionObject))
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Its purely service REST contract and these concerns should not be in SDK. Please move to friends.

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: How do you want to move it to friends without access to the class members?

**Pull request merged**